File: packages/js/src/decompose.ts
import type { UnitDef } from './types'
import type { UnitRegistry } from './units'
import type { NormalizedUsage } from './usage'

import { getActiveRegistry, isDescendantOrSelf } from './units'
import { getUsageValue } from './usage'

export function computeLeafValues(
  pricedUsageKeys: Set<string>,
  usage: NormalizedUsage,
  registry: UnitRegistry = getActiveRegistry()
): Record<string, number> {
  const pricedUnits = [...pricedUsageKeys]
    .map((usageKey) => registry.getUnit(usageKey))
    .filter((unit): unit is UnitDef => unit !== undefined)
    .sort(
      (left, right) =>
        Object.keys(right.dimensions).length - Object.keys(left.dimensions).length || left.usageKey.localeCompare(right.usageKey)
    )
  const leafValues: Record<string, number> = {}

  for (const unit of pricedUnits) {
    const unitValue = getUsageValue(usage, unit.usageKey, registry)
    let descendantLeafTotal = 0
    for (const descendant of pricedUnits) {
      if (descendant === unit || !isDescendantOrSelf(unit, descendant)) continue

      const descendantLeafValue = leafValues[descendant.usageKey]
      if (descendantLeafValue === undefined) {
        throw new Error(`Missing computed leaf value for ${descendant.usageKey}`)
      }
      descendantLeafTotal += descendantLeafValue
    }

    let leafValue = unitValue - descendantLeafTotal
    const roundingTolerance = Number.EPSILON * Math.max(1, Math.abs(unitValue), Math.abs(descendantLeafTotal)) * (pricedUnits.length + 1)
    if (leafValue < 0 && Math.abs(leafValue) <= roundingTolerance) {
      leafValue = 0
    }

    if (leafValue < 0) {
      throw new Error(negativeLeafErrorMessage(unit, pricedUnits, usage, leafValue, registry))
    }

    leafValues[unit.usageKey] = leafValue
  }

  return leafValues
}

function negativeLeafErrorMessage(
  unit: UnitDef,
  pricedUnits: UnitDef[],
  usage: NormalizedUsage,
  leafValue: number,
  registry: UnitRegistry
): string {
  const unitValue = getUsageValue(usage, unit.usageKey, registry)
  const descendantValues = pricedUnits
    .filter((descendant) => descendant !== unit && isDescendantOrSelf(unit, descendant))
    .map((descendant) => ({
      unit: descendant,
      value: getUsageValue(usage, descendant.usageKey, registry),
    }))
    .filter(({ value }) => value > 0)

  for (const descendant of descendantValues) {
    if (descendant.value > unitValue) {
      return `Invalid usage data: ${descendant.unit.usageKey} (${descendant.value.toString()}) cannot exceed ${unit.usageKey} (${unitValue.toString()})`
    }
  }

  const descendantKeys = descendantValues.map(({ unit }) => unit.usageKey).join(', ')
  const descendantTotal = unitValue - leafValue
  return `Invalid usage data: more-specific usage for ${descendantKeys} totals ${descendantTotal.toString()}, which exceeds ${unit.usageKey} (${unitValue.toString()})`
}

File: packages/python/genai_prices/decompose.py
from __future__ import annotations

from collections.abc import Mapping
from typing import cast

from genai_prices._usage import UsageValue, subtract_usage_values
from genai_prices.units import UnitDef


def is_descendant_or_self(ancestor: UnitDef, descendant: UnitDef) -> bool:
    return ancestor.dimensions.items() <= descendant.dimensions.items()


def compute_leaf_values(
    priced_usage_keys: set[str], usage: object, units_by_usage_key: Mapping[str, UnitDef]
) -> dict[str, UsageValue]:
    priced_units = sorted(
        (units_by_usage_key[usage_key] for usage_key in priced_usage_keys & units_by_usage_key.keys()),
        key=lambda unit: (-len(unit.dimensions), unit.usage_key),
    )
    leaf_values: dict[str, UsageValue] = {}

    for unit in priced_units:
        descendant_values = (
            leaf_values[descendant.usage_key]
            for descendant in priced_units
            if descendant is not unit and is_descendant_or_self(unit, descendant)
        )
        leaf_value = subtract_usage_values(_usage_value(usage, unit.usage_key), descendant_values)

        if leaf_value < 0:
            raise ValueError(_negative_leaf_error_message(unit, priced_units, usage, leaf_value))

        leaf_values[unit.usage_key] = leaf_value

    return leaf_values


def _usage_value(usage: object, usage_key: str) -> UsageValue:
    raw_value = getattr(usage, usage_key, None)
    return 0 if raw_value is None else cast(UsageValue, raw_value)


def _negative_leaf_error_message(
    unit: UnitDef, priced_units: list[UnitDef], usage: object, leaf_value: UsageValue
) -> str:
    unit_value = _usage_value(usage, unit.usage_key)
    descendant_values = [
        (descendant, value)
        for descendant in priced_units
        if descendant is not unit
        and is_descendant_or_self(unit, descendant)
        and (value := _usage_value(usage, descendant.usage_key)) > 0
    ]

    for descendant, value in descendant_values:
        if value > unit_value:
            return f'Invalid usage data: {descendant.usage_key} ({value}) cannot exceed {unit.usage_key} ({unit_value})'

    descendant_keys = ', '.join(descendant.usage_key for descendant, _ in descendant_values)
    descendant_total = subtract_usage_values(unit_value, (leaf_value,))
    return (
        f'Invalid usage data: more-specific usage for {descendant_keys} totals {descendant_total}, '
        f'which exceeds {unit.usage_key} ({unit_value})'
    )
